archinfo/arch_soot.py

```diff
@@ -123,25 +171,145 @@ def __repr__(self):
+        return self.name
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __eq__(self, other):
+        return isinstance(other, SootClassDescriptor) and \
+               self.name == other.name
+
+    def __ne__(self, other):
+        return not self == other
+
+    @property
+    def is_loaded(self):
+        """
+        :return: True, if the class is loaded in CLE and thus
+                 info about field, methods, ... are available.
+        """
+        return self._soot_class != None
+
+    @property
+    def fields(self):
+        return self._soot_class.fields if self.is_loaded else None
+  
+    @property
+    def methods(self):
+        return self._soot_class.methods if self.is_loaded else None
+    
+    @property
+    def superclass_name(self):
+        return self._soot_class.super_class if self.is_loaded else None
+
+
+class ArchSoot(Arch):
+    def __init__(self, endness=Endness.LE):
+        super(ArchSoot, self).__init__(endness)
+
+    name = 'Soot'
 
     vex_arch = None  # No VEX support
     qemu_name = None  # No Qemu/Unicorn-engine support
     bits = 64
     address_types = (SootAddressDescriptor, )
     function_address_types = (SootMethodDescriptor, )
 
-    name = 'Soot'
+    # Size of native counterparts of primitive Java types
+    sizeof = {'boolean'  :  8,
+              'byte'     :  8,
+              'char'     : 16, 
+              'short'    : 16,
+              'int'      : 32,
+              'long'     : 64, 
+              'float'    : 32,
+              'double'   : 64
+             }
+
+    primitive_types = ['boolean',
+                       'byte',
+                       'char',
+                       'short',
+                       'int',
+                       'long',
+                       'float',
+                       'double']
+
+    sig_dict = {'Z' : 'boolean',
+                'B' : 'byte',
+                'C' : 'char',
+                'S' : 'short',
+                'I' : 'int',
+                'J' : 'long',
+                'F' : 'float',
+                'D' : 'double',
+                'V' : 'void'
+               } 
+
+    @staticmethod
+    def decode_type_signature(type_sig):
+
+        if type_sig == "":
+            return None
+
+        if type_sig in ArchSoot.sig_dict:
+            return ArchSoot.sig_dict[type_sig]
+
+        if type_sig.startswith('L') and type_sig.endswith(';'):
+            return type_sig[1:-1]
+
+        raise ValueError(type_sig)
+
+    @staticmethod
+    def decode_method_signature(method_sig):
+        match = re.search(r'\((.*)\)(.*)', method_sig)
+        
+        params = tuple(ArchSoot.decode_type_signature(param) 
+                       for param in re.findall(r'([\[]*[ZBCSIJFDV]|[\[]*L.+;)', match.group(1)))
+        ret = ArchSoot.decode_type_signature(match.group(2))
+
+        l.debug("Decoded method signature '%s' as params=%s and ret=%s"
+                "" % (method_sig, params, ret) )
```

Comment:
Do not format strings in `.debug()` calls. You should rewrite it as the following:

```
l.debug("Decoded method signature '%s' as params=%s and ret=%s",
        method_sig, params, ret
        )
```